### PR TITLE
Allow for customizable ModalRoute barrierTween

### PR DESCRIPTION
**Pull request by shihaohong**
## Description

This allows you to customize the barrier tween for any ModalRoute. It currently only defaults to `Curves.ease`. This allows for any subclass of ModalRoute to implement its own `barrierCurve`.

```dart
class _DialogRouteWithCustomBarrierCurve<T> extends PopupRoute<T> {
  _DialogRouteWithCustomBarrierCurve({
    Animatable<double> barrierCurve,
    // ...
  }) : assert(barrierCurve!= null),
       // ...
       _barrierCurve = barrierCurve,
  // ...

  @override
  Animatable<double> get barrierCurve => barrierCurve;
  final Animatable<double> _barrierCurve;

  // ...
```

## Related Issues

n/a

## Tests

I added the following tests:

- A test to validate that the default barrier tween remains as `CurveTween(curve: Curves.ease)`
- A test to validate that a custom barrier tween is honored.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/material/page.dart [hunk 1]
```diff
@@ -44,11 +44,14 @@ class MaterialPageRoute<T> extends PageRoute<T> {
     @required this.builder,
     RouteSettings settings,
     this.maintainState = true,
+    Duration transitionDuration = const Duration(milliseconds: 300),
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/page.dart]: Are the changes in this file supposed to be here? They seem unrelated...

File: packages/flutter/lib/src/widgets/routes.dart [hunk 2]
```diff
@@ -1060,6 +1060,31 @@ abstract class ModalRoute<T> extends TransitionRoute<T> with LocalHistoryRoute<T
...
+  /// generally prevents the user from interacting with the route below the
+  /// current route, and normally partially obscures such routes.
+  ///
+  /// For example, when a dialog is on the screen, the page below the dialog is
+  /// usually darkened by the modal barrier.
+  ///
+  /// While the route is animating into position, the color is animated from
+  /// transparent to the specified color.
+  ///
+  /// If this getter would ever start returning a different color,
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/routes.dart]: This sentence is odd, the getter doesn't return a color?

File: packages/flutter/lib/src/widgets/routes.dart [hunk 3]
```diff
@@ -1060,6 +1060,31 @@ abstract class ModalRoute<T> extends TransitionRoute<T> with LocalHistoryRoute<T
...
+  /// effect.
+  ///
+  /// It defaults to a [CurveTween] with a [Curves.ease] curve.
+  ///
+  /// See also:
+  ///
+  ///  * [barrierColor], which determines the color that the modal transitions
+  ///    to.
+  ///  * [AnimatedModalBarrier], the widget that implements this feature.
+  Animatable<double> get barrierTween => CurveTween(curve: Curves.ease);
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/routes.dart]: This is kinda odd since the full barrierTween would really also include the ColorTween? I wonder what you would use this except for changing the curve? Should we just make the curve customizable? Or are there other use cases?

File: packages/flutter/lib/src/material/page.dart (hunk 1 above)
Comment (shihaohong) [reply packages/flutter/lib/src/material/page.dart]: Oops, these are unrelated and are part of something else I'm working on. I'll remove these.

File: packages/flutter/lib/src/widgets/routes.dart (hunk 2 above)
Comment (shihaohong) [reply packages/flutter/lib/src/widgets/routes.dart]: Sorry, this was a remnant of copying the docs from the rest of the `barrier__` properties. I notice that it's the same way for the non-`barrierColor` properties and updated them in those places since those didn't make sense either.

File: packages/flutter/lib/src/widgets/routes.dart (hunk 3 above)
Comment (shihaohong) [reply packages/flutter/lib/src/widgets/routes.dart]: It becomes sort of weird since currently, the internal `ColorTween`  uses the `barrierColor` as an input. If we have `barrierTween` be the entire customizable curve, then it could theoretically ignore `barrierColor`. An example would be:
```dart
PopupRoute(
  barrierTween: ColorTween(
    begin: _kTransparent,
    end: Colors.orange, // no dependence on barrierColor
  ).chain(CurveTween(curve: Curves.ease));
  // ...
);
``` 
Would that be worth making the entire curve customizable?

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/routes.dart]: I was thinking we limit this even more. Instead of allowing you to customize the CurveTween you can only customize the Curve used by that CurveTween.

**Comment by shihaohong** [comment]
@goderbauer Should be ready for another round of review. PTAL

File: packages/flutter/lib/src/widgets/routes.dart [hunk 4]
```diff
@@ -1060,6 +1060,32 @@ abstract class ModalRoute<T> extends TransitionRoute<T> with LocalHistoryRoute<T
...
+  ///
+  /// The modal barrier is the scrim that is rendered behind each route, which
+  /// generally prevents the user from interacting with the route below the
+  /// current route, and normally partially obscures such routes.
+  ///
+  /// For example, when a dialog is on the screen, the page below the dialog is
+  /// usually darkened by the modal barrier.
+  ///
+  /// While the route is animating into position, the color is animated from
+  /// transparent to the specified color.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/routes.dart]:
```suggestion
  /// transparent to the specified [barrierColor].
```

File: packages/flutter/lib/src/widgets/routes.dart [hunk 5]
```diff
@@ -1060,6 +1060,32 @@ abstract class ModalRoute<T> extends TransitionRoute<T> with LocalHistoryRoute<T
...
+  /// [changedInternalState] should be invoked so that the change can take
+  /// effect.
+  ///
+  /// It defaults to [Curves.ease].
+  ///
+  /// See also:
+  ///
+  ///  * [barrierColor], which determines the color that the modal transitions
+  ///    to.
+  ///  * [Curves] for collection of common curves.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/routes.dart]:
```suggestion
  ///  * [Curves] for a collection of common curves.
```

File: packages/flutter/test/widgets/routes_test.dart [hunk 6]
```diff
@@ -864,6 +864,133 @@ void main() {
...
+                ),
+              );
+            }
+          ),
+        ),
+      ));
+
+      final CurveTween _defaultBarrierTween = CurveTween(curve: Curves.ease);
+      double _getBarrierTweenAlphaValue(double t) {
+        return _defaultBarrierTween.transform(t) * 255;
```
Comment (goderbauer) [inline packages/flutter/test/widgets/routes_test.dart]: Use https://master-api.flutter.dev/flutter/dart-ui/Color/getAlphaFromOpacity.html to convert from opacity to Alpha?

File: packages/flutter/test/widgets/routes_test.dart [hunk 7]
```diff
@@ -864,6 +864,133 @@ void main() {
...
+                  },
+                ),
+              );
+            }
+          ),
+        ),
+      ));
+
+      final CurveTween _defaultBarrierTween = CurveTween(curve: Curves.ease);
+      double _getBarrierTweenAlphaValue(double t) {
```
Comment (goderbauer) [inline packages/flutter/test/widgets/routes_test.dart]: Maybe rename to "getExpectedBarrierAlphaValue"?

(I was confused because this didn't seem to access the actual value from the barrier).

File: packages/flutter/test/widgets/routes_test.dart [hunk 8]
```diff
@@ -864,6 +864,133 @@ void main() {
...
+                    );
+                  },
+                ),
+              );
+            },
+          ),
+        ),
+      ));
+
+      final CurveTween _customBarrierTween = CurveTween(curve: Curves.linear);
```
Comment (goderbauer) [inline packages/flutter/test/widgets/routes_test.dart]: Same comments here.

File: packages/flutter/test/widgets/routes_test.dart [hunk 9]
```diff
@@ -892,3 +1019,68 @@ class DialogObserver extends NavigatorObserver {
     super.didPush(route, previousRoute);
   }
 }
+
+class _TestDialogRoute<T> extends PopupRoute<T> {
```
Comment (goderbauer) [inline packages/flutter/test/widgets/routes_test.dart]: You could potentially fold this and the next class into one by just calling "super.barrierCurve" if the barrierCurve provided in the constructor was null. It's up to you though.

**Review by goderbauer — APPROVED** [review]
LGTM
